audio/out/ao_sndio.c

```diff
@@ -0,0 +1,330 @@
+    pthread_mutex_t lock;
+    /* data control */
+    void *buffer;
+    int delay;
+    /* volume control */
+    int vol;
+    bool havevol;
+};
+
+/*
+ * misc parameters (volume, etc...)
+ */
+static int control(struct ao *ao, enum aocontrol cmd, void *arg)
+{
+    struct priv *p = ao->priv;
+    ao_control_vol_t *vol = arg;
+
+    switch (cmd) {
+    case AOCONTROL_GET_VOLUME:
+        if (!p->havevol)
+            return CONTROL_FALSE;
+        vol->left = vol->right = p->vol * 100 / (float)SIO_MAXVOL;
+        break;
+    case AOCONTROL_SET_VOLUME:
+        if (!p->havevol)
+            return CONTROL_FALSE;
+        /* hold the lock while touching p->hdl */
+        pthread_mutex_lock(&p->lock);
+        sio_setvol(p->hdl, vol->left * SIO_MAXVOL / 100);
+        pthread_mutex_unlock(&p->lock);
+        break;
+    default:
+        return CONTROL_UNKNOWN;
+    }
+    return CONTROL_OK;
+}
+
+/*
+ * call-back invoked to notify of the hardware position
+ */
+static void movecb(void *addr, int delta)
+{
+    struct priv *p = addr;
+    p->delay -= delta;
+}
+
+/*
+ * call-back invoked to notify about volume changes
+ */
+static void volcb(void *addr, unsigned newvol)
+{
+    struct priv *p = addr;
+    p->vol = newvol;
+}
+
+static const struct mp_chmap sndio_layouts[MP_NUM_CHANNELS + 1] = {
+    {0},                                        // empty
+    MP_CHMAP_INIT_MONO,                         // mono
+    MP_CHMAP_INIT_STEREO,                       // stereo
+    {0},                                        // 2.1
+    MP_CHMAP4(FL, FR, BL, BR),                  // 4.0
+    {0},                                        // 5.0
+    MP_CHMAP6(FL, FR, BL, BR, FC, LFE),         // 5.1
+    {0},                                        // 6.1
+    MP_CHMAP8(FL, FR, BL, BR, FC, LFE, SL, SR), // 7.1
+    /* above is the fixed channel assignment for sndio, since we need to fill
+       all channels and cannot insert silence, not all layouts are supported.
+       NOTE: MP_SPEAKER_ID_NA could be used to add padding channels. */
+};
+
+/*
+ * open device and setup parameters
+ * return: 0=success -1=fail
+ */
+static int init(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    struct af_to_par {
+        int format, bits, sig;
+    };
+    static const struct af_to_par af_to_par[] = {
+        {AF_FORMAT_U8,   8, 0},
+        {AF_FORMAT_S16, 16, 1},
+        {AF_FORMAT_S32, 32, 1},
+    };
+    const struct af_to_par *ap;
+    int i;
+
+    /* support audio device defined in the command line */
+    const char *dev_name = ao->device ? ao->device : SIO_DEVANY;
+    /* use blocking mode */
+    p->hdl = sio_open(dev_name, SIO_PLAY, 0);
+    if (p->hdl == NULL) {
+        MP_ERR(ao, "can't open sndio device '%s'\n", dev_name);
+        goto error;
+    }
+
+    ao->format = af_fmt_from_planar(ao->format);
+
+    sio_initpar(&p->par);
+    for (i = 0, ap = af_to_par;; i++, ap++) {
+        if (i == sizeof(af_to_par) / sizeof(struct af_to_par)) {
```

Comment:
That's a bit cleaner, but I think

```
            if (ap->bits > 8)
                p->par.le = SIO_LE_NATIVE;
```

is unnecessary? Since it's set to `SIO_LE_NATIVE` by default.

The 

```
 if (ap->bits != SIO_BPS(ap->bits))
                p->par.bps = ap->bits / 8;
```

is weird to me as well, since it's comparing a number of bytes with a number of bits.

Reply:
That last block was in the original code as well, which is quite interesting.